anndata/tests/test_dask.py

```diff
@@ -20,3 +21,88 @@ def test_dask_X_view():
     adata.X = da.ones((M, N))
     view = adata[:30]
     view.copy()
+
+
+# TODO: Appropriate path to write in testing.
+# TODO: Check written object
+# TODO: Check that this works with `zarr`
+
+
+def test_dask_write_h5ad():
+    import dask.array as da
+
+    M, N = 50, 30
+    adata = ad.AnnData(
+        obs=pd.DataFrame(index=[f"cell{i:02}" for i in range(M)]),
+        var=pd.DataFrame(index=[f"gene{i:02}" for i in range(N)]),
+    )
+    adata.X = da.ones((M, N))
+    view = adata[:30]
+    view.copy()
+    view.write_h5ad("test.h5ad")
+
+
+# TODO: Appropriate path to write in testing.
+# TODO: Check written object
+
+
+def test_dask_write():
+    import dask.array as da
+
+    M, N = 50, 30
+    adata = ad.AnnData(
+        obs=pd.DataFrame(index=[f"cell{i:02}" for i in range(M)]),
+        var=pd.DataFrame(index=[f"gene{i:02}" for i in range(N)]),
+    )
+    adata.X = da.ones((M, N))
+    view = adata[:30]
+    view.copy()
+    view.write("test.h5ad")
```

Comment:
I would suggest something a little more like this.

```suggestion
@pytest.fixture(params=["h5ad", "zarr"])
def diskfmt(request):
    return request.param


def test_dask_write(tmp_path, diskfmt):
    import dask.array as da

    pth = tmp_path / f"test_write.{diskfmt}"
    write = lambda x, y: getattr(x, f"write_{diskfmt}")(y)
    read = lambda x: getattr(ad, f"read_{diskfmt}")(x)

    M, N = 50, 30
    adata = ad.AnnData(
        obs=pd.DataFrame(index=[f"cell{i:02}" for i in range(M)]),
        var=pd.DataFrame(index=[f"gene{i:02}" for i in range(N)]),
    )
    adata.X = da.random.randint((M, N))
    write(adata, pth)
    result = read(pth)

    assert_equal(adata, result)
```

* The diskfmt fixture lets us test both the hdf5 and zarr backends
* We're checking that our result is equal to what we expect
* I would have one test for views, one which doesn't use views.

Reply:
After applying your suggestions, I realized making these changes makes the tests wrong indeed. But I modified the write tests, and I was surprised that they passed. Turns out, AnnData converts X to ndarrays before storing.
First I will focus on the `test_dask_write` cases which gives not supported write error when obsm/varm gets assigned.

And in the future for X, we could first focus on this case:
``` python
assert type(adata.X) == da.Array # doesn't pass
```